tests/integration/test_client_http.py

```diff
@@ -0,0 +1,203 @@
+        try:
+            client.describe_regions()
+        except EndpointConnectionError:
+            pass
+
+    def test_bad_status_line(self):
+        config = Config(retries={'max_attempts': 0})
+        client = self.session.create_client('ec2', endpoint_url=self.localhost,
+                                            config=config)
+
+        class BadStatusHandler(BaseHTTPServer.BaseHTTPRequestHandler):
+            def do_POST(self):
+                self.wfile.write(b'garbage')
+
+        try:
+            with background(run_server, args=(BadStatusHandler, self.port)):
+                client.describe_regions()
+        except ConnectionClosedError:
+            pass
+        except BackgroundTaskFailed:
+            self.fail('Fake EC2 service was not called.')
+        else:
+            self.fail('Excepted exception was not thrown')
+
+
+def random_port():
+    return random.randint(8080, 9000)
+
+
+class SimpleHandler(BaseHTTPServer.BaseHTTPRequestHandler):
+    status = 200
+
+    def get_length(self):
+        return 0
+
+    def get_body(self):
+        return b''
+
+    def do_GET(self):
+        length = str(self.get_length())
+        self.send_response(self.status)
+        self.send_header('Content-Length', length)
+        self.end_headers()
+        self.wfile.write(self.get_body())
+
+    do_POST = do_PUT = do_GET
+
+
+class ProxyHandler(BaseHTTPServer.BaseHTTPRequestHandler):
+    tunnel_chunk_size = 1024
+    def _tunnel(self, client, remote):
+        client.setblocking(0)
+        remote.setblocking(0)
+        sockets = [client, remote]
+        while True:
+            readable, writeable, _ = select.select(sockets, sockets, [], 1)
+            if client in readable and remote in writeable:
+                client_bytes = client.recv(self.tunnel_chunk_size)
+                if not client_bytes:
+                    break
+                remote.sendall(client_bytes)
+            if remote in readable and client in writeable:
+                remote_bytes = remote.recv(self.tunnel_chunk_size)
+                if not remote_bytes:
+                    break
+                client.sendall(remote_bytes)
+
+    def do_CONNECT(self):
+        if not self.validate_auth():
+            self.send_response(401)
+            self.end_headers()
+            return
+
+        self.send_response(200)
+        self.end_headers()
+
+        remote_host, remote_port = self.path.split(':')
+        remote_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        remote_socket.connect((remote_host, int(remote_port)))
+
+        self._tunnel(self.request, remote_socket)
+        remote_socket.close()
+
+    def validate_auth(self):
+        return True
+
+
+class BackgroundTaskFailed(Exception):
+    pass
+
+
+@contextmanager
+def background(target, args=(), timeout=10):
+    thread = threading.Thread(target=target, args=args)
+    thread.daemon = True
+    thread.start()
+    yield target
```

Comment:
You'll need a try/finally here.  For the tests that are expecting a failure being raised, the code after the yield won't be called.  It doesn't technically matter because they are background threads, but it's a pretty small change to fix this.

Reply:
Done